common/crypto/src/hkdf.rs

```diff
@@ -28,3 +31,64 @@ where
 
     Ok(okm)
 }
+
+/// `DerivationMaterial` encapsulates parameters for deterministic key derivation using
+/// HKDF (SHA-512).
+///
+/// It consists of:
+///   - A master key (`master_key`): the base secret.
+///   - An index (`index`): ensures unique derivations.
+///   - A salt (`salt`): adds additional uniqueness.
+///
+/// Use the `derive_secret()` method to generate a 32-byte secret. To prepare for a new derivation,
+/// call the `next()` method, which increments the index. **It is the caller's responsibility to
+/// track and persist the derivation index if keys need to be rederived.**
+///
+/// # Example
+///
+/// ```rust
+/// let master_key = [0u8; 32]; // your secret master key
+/// let salt = "unique-salt-value".to_string();
+/// let material = DerivationMaterial::new(master_key, 0, salt);
+///
+/// // Derive a secret
+/// let secret = material.derive_secret().expect("Failed to derive secret");
+///
+/// // Prepare for the next derivation
+/// let next_material = material.next();
+/// ```
+pub struct DerivationMaterial {
+    master_key: [u8; 32],
+    index: u32,
+    salt: String,
+}
+
+impl DerivationMaterial {
+    /// Derives a 32-byte seed from a master seed and an index using HKDF (with SHA-512).
+    ///
+    /// The `salt` and the use of the index (as info) bind this derivation to an application/client.
+    pub fn derive_secret(&self) -> Result<[u8; 32], hkdf::InvalidLength> {
```

Comment:
also it seems like we might be reinventing the wheel here a bit, aren't there some libraries for this purpose out there? something like bip32 maybe?